qiskit/providers/experiment/analysis_result.py

```diff
@@ -0,0 +1,377 @@
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Analysis result abstract interface."""
+
+import logging
+from typing import Optional, List, Union, Dict, Callable, Any
+import uuid
+import json
+import copy
+from functools import wraps
+
+from .constants import ResultQuality
+from .json import NumpyEncoder, NumpyDecoder
+from .utils import save_data, qiskit_version
+from .exceptions import ExperimentError
+from .device_component import DeviceComponent, to_component
+
+LOG = logging.getLogger(__name__)
+
+
+def auto_save(func: Callable):
+    """Decorate the input function to auto save data."""
+
+    @wraps(func)
+    def _wrapped(self, *args, **kwargs):
+        return_val = func(self, *args, **kwargs)
+        if self.auto_save:
+            self.save()
+        return return_val
+
+    return _wrapped
+
+
+class AnalysisResult:
+    """Base common type for all versioned AnalysisResult abstract classes.
+
+    Note this class should not be inherited from directly, it is intended
+    to be used for type checking. When implementing a provider you should use
+    the versioned abstract classes as the parent class and not this class
+    directly.
+    """
+
+    version = 0
+
+
+class AnalysisResultV1(AnalysisResult):
+    """Class representing an analysis result for an experiment."""
+
+    version = 1
+    _data_version = 1
+
+    _json_encoder = NumpyEncoder
+    _json_decoder = NumpyDecoder
+
+    _extra_data = {}
+
+    def __init__(
+        self,
+        result_data: Dict,
+        result_type: str,
+        device_components: List[Union[DeviceComponent, str]],
+        experiment_id: str,
+        result_id: Optional[str] = None,
+        quality: Union[ResultQuality, str] = ResultQuality.UNKNOWN,
+        verified: bool = False,
+        tags: Optional[List[str]] = None,
+        service: Optional["ExperimentServiceV1"] = None,
+        **kwargs,
+    ):
+        """AnalysisResult constructor.
+
+        Args:
+            result_data: Analysis result data.
+            result_type: Analysis result type.
+            device_components: Target device components this analysis is for.
+            experiment_id: ID of the experiment.
+            result_id: Result ID. If ``None``, one is generated.
+            quality: Quality of the analysis.
+            verified: Whether the result quality has been verified.
+            tags: Tags for this analysis result.
+            service: Experiment service to be used to store result in database.
+            **kwargs: Additional analysis result attributes.
+        """
+        result_data = result_data or {}
+        self._result_data = copy.deepcopy(result_data)
+        self._source = self._result_data.pop(
```

Comment:
Are you sure that it's `pop` and not `get` here?

Reply:
Yeah because `_source` is an "internal" metadata and not part of the actual result data. It's stored in the result data dictionary in the database because there's no other place to store it, but it's removed from the data dictionary when the AnalysisResult object is loaded from the database.